lib/msal-common/src/request/AuthenticationHeaderParser.ts

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ClientConfigurationError } from "..";
+import { HeaderNames } from "../utils/Constants";
+
+type WWWAuthenticateChallenges = {
+    nonce?: string,
+};
+
+type AuthenticationInfoChallenges = {
+    nextnonce?: string
+};
+
+/**
+ * This is a helper class that parses supported HTTP response authentication headers to extract and return
+ * header challenge values that can be used outside the basic authorization flows.
+ */
+export class AuthenticationHeaderParser{
+    private headers: Headers;
+
+    constructor(headers: Headers) {
+        this.headers = headers;
+    }
+
+    /**
+     * This method parses the SHR nonce value out of either the Authentication-Info or WWW-Authenticate authentication headers.
+     * @returns 
+     */
+    getShrNonce(): string {
+        // Attempt to parse nonce from Authentiacation-Info
+        const authenticationInfo = this.headers.get(HeaderNames.AuthenticationInfo);
+        if (authenticationInfo) {
+            const authenticationInfoChallenges = this.parseChallenges(authenticationInfo) as AuthenticationInfoChallenges;
+            if (authenticationInfoChallenges.nextnonce) {
+                return authenticationInfoChallenges.nextnonce;
+            }
+            throw ClientConfigurationError.createInvalidAuthenticationHeaderError(HeaderNames.AuthenticationInfo, "nextnonce challenge is missing.");
+        }
+
+        // Attempt to parse nonce from WWW-Authenticate
+        const wwwAuthenticate = this.headers.get(HeaderNames.WWWAuthenticate);
+        if (wwwAuthenticate) {
+            const wwwAuthenticateChallenges = this.parseChallenges(wwwAuthenticate) as WWWAuthenticateChallenges;        
```

Comment:
Looks like `parseChallenages` can be made into a generic.

Reply:
Done